services/business/checkers/destination_rules/multi_match_checker.go

```diff
@@ -0,0 +1,103 @@
+package destination_rules
+
+import (
+	"strings"
+
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/services/models"
+)
+
+type MultiMatchChecker struct {
+	DestinationRules []kubernetes.IstioObject
+}
+
+type Host struct {
+	Service   string
+	Namespace string
+	Cluster   string
+}
+
+func (m MultiMatchChecker) Check() models.IstioValidations {
+	validations := models.IstioValidations{}
+
+	var empty struct{}
+	seenHosts := make(map[string]map[string]map[string]struct{}) // Poor man's trie
```

Comment:
For your comment at the end of the line, I guess there is something you don't like that much in this type. Probably lack of readiness.
What do you think about moving this type to sub-types? I could help other's devs to better understand what we are doing here. Besides that, I have the feeling that this code will be reused by other components.
I was thinking of something like the following. Perhaps, different naming and without the string types.

```golang
type Cluster string
type Namespace string
type Service string
type ClusterSearchMap map[Cluster]NamespaceSearchMap
type NamespaceSearchMap map[Namespace]ServiceSearchMap
type ServiceSearchMap map[Service]struct{}
```

Reply:
Readability is one, yes. The other thing is that it's pretty limited structure. But as long as we can guarantee that there's no need for other query types than these, then it's going to be fine. But what if we need something like ``*-frontend.namespace.cluster`` and the ``*-frontend`` could match ``service1-frontend`` as well as ``service1-qa-frontend``. That's when we would need a better sorted data structure (or other prefix-searchable)